watchevents/namespaces.go

```diff
@@ -0,0 +1,95 @@
+package watchevents
+
+import (
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/kubernetes-agent/kubernetesclient"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+type namespaceHandler struct {
+	rClient     *client.RancherClient
+	kClient     *kubernetesclient.Client
+	nsWatchChan chan struct{}
+}
+
+func NewNamespaceHandler(rClient *client.RancherClient, kClient *kubernetesclient.Client) *namespaceHandler {
+	nsHandler := &namespaceHandler{
+		rClient: rClient,
+		kClient: kClient,
+	}
+	return nsHandler
+}
+
+func (n *namespaceHandler) startNamespaceWatch() chan struct{} {
+	watchlist := cache.NewListWatchFromClient(n.kClient.K8sClient.Core().RESTClient(), "namespaces", "", fields.Everything())
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Namespace{},
+		time.Second*0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
```

Comment:
You should define concrete types in func() instead of obj interface{}. And pass those concrete types to corresponding handlers like n.delete. The same applies to all informers